### PR TITLE
layout: Implement `baseline-source` 

### LINKED ISSUE #31723: Enable the internal -x-lang CSS property
The behavior of `text-transform` can be affected by the language of the text, e.g.

```html
<div style="text-transform: uppercase">
  <span lang="tr">i</span> <!-- İ -->
  <span lang="en">i</span> <!-- I -->
</div>
```

Therefore, we need some way to track the language.

Gecko uses an internal `-x-lang` CSS property for that, so it seems a good idea to do the same (especially since we use the same style engine).

To do that,

 1. Clone a fork of https://github.com/servo/stylo/
 1. Enable `-x-lang` not just for `gecko` but also `servo`.
   https://github.com/servo/stylo/blob/7dd884031d6b97d801d50f8fcdff69bbeb292ae0/style/properties/longhands/font.mako.rs#L257-L266
 1. Run `cargo build --features servo` to ensure it builds fine
 1. Open a PR at https://github.com/servo/stylo/
 1. Here in Servo, update `Cargo.toml` to take the various style-related crates from your PR. Compile, this will automatically update `Cargo.lock`. See 871e36f5a5f8d87cd8d3979794e8e21588f54939 for reference.
 1. In this case I don't expect anything to break, but otherwise you would need to do the changes necessary so that it still compiles fine.
 1. Open a PR with the Servo changes, so that we can run the tests and ensure everything is fine. `-x-lang` is an internal property, so I don't expect any test to complain.
 1. Then we can land the PR for Stylo
 1. Then update the PR for Servo, reverting the changes from `Cargo.toml`, but bumping `Cargo.lock` to point to the new upstream rev. See https://github.com/servo/servo/commit/0935bf52fd17e05c0ef4f0a2ef560dca32a4fa23 for reference.
 1. Finally we can land the PR for Servo

### LINKED ISSUE #31875: Implement baseline-source
After #31705 I think it should be quite straightforward to implement `baseline-source`: https://drafts.csswg.org/css-inline/#baseline-source

This will require Stylo changes, so you can follow the instructions from #31723

Basically https://github.com/servo/stylo/blob/bcdebc41b2a2222e9fa5ae6d7d1572025c32dc47/style/properties/longhands/box.mako.rs#L107 needs to change the obsolete `servo-2013` into `servo`

And then https://github.com/servo/servo/blob/d814d0553981777c01409d545e5c1c335b8eb7f4/components/layout_2020/flow/inline.rs#L2145-L2153 needs to check `BaselineSource`, picking the 1st baseline for `BaselineSource::First`, the last baseline for `BaselineSource::Last`, and keep the current logic for `BaselineSource::Auto`.

### PR DESCRIPTION
**Pull request by MunishMummadi**
**Description**
-Enhanced the `pick_baseline` function to dynamically select the baseline according to the `BaselineSource` property (First, Last, Auto), allowing for more precise alignment of inline elements based on CSS specifications.

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #31875 

addresses the  [issue](https://github.com/servo/servo/issues/31875)

### CODE REVIEW

**Review by Loirooriol — COMMENTED** [review]
This is looking good, but no test passes because they are in `css/css-inline/baseline-source/`, but we are skipping `css/css-inline/` in general: https://github.com/servo/servo/blob/da76ebabe9359c6f815f16adc81af7f11bb3265c/tests/wpt/include.ini#L50-L51

So can you create another PR enabling the baseline-source tests? Then here we will see how many improve.

File: components/layout_2020/flow/inline.rs [hunk 1]
```diff
@@ -2143,13 +2144,20 @@ impl IndependentFormattingContext {
     /// Picks either the first or the last baseline, depending on `baseline-source`.
     /// <https://drafts.csswg.org/css-inline/#baseline-source>
     fn pick_baseline(&self, baselines: &Baselines) -> Option<Au> {
-        // TODO: Currently this only supports the initial `baseline-source: auto`.
-        if let Self::NonReplaced(non_replaced) = self {
-            if let NonReplacedFormattingContextContents::Flow(_) = non_replaced.contents {
-                return baselines.last;
-            }
+        // Checking the baseline source
```
[the remark this answers is not in the document]
Comment (MunishMummadi) [reply components/layout_2020/flow/inline.rs]: removed it!!

**Comment by MunishMummadi** [comment]
Hey @Loirooriol Sorry . I made changes and committed to this PR. Any way I can resolve. I can push a new PR without `[css-inline/baseline-source]
skip: false`(if you want).

**Comment by Loirooriol** [comment]
You will need to add various tests expectations once they aren't skipped, so it's better to do it in another PR to avoid noise here.

File: components/layout_2020/flow/inline.rs [hunk 2]
```diff
@@ -2143,13 +2144,19 @@ impl IndependentFormattingContext {
...
-        // TODO: Currently this only supports the initial `baseline-source: auto`.
-        if let Self::NonReplaced(non_replaced) = self {
-            if let NonReplacedFormattingContextContents::Flow(_) = non_replaced.contents {
-                return baselines.last;
-            }
+        match self.style().clone_baseline_source() {
+            BaselineSource::First => baselines.first,
+            BaselineSource::Last => baselines.last,
+            BaselineSource::Auto => {
+                // The existing logic for `Auto` remains unchanged.
```
Comment (Loirooriol) [inline components/layout_2020/flow/inline.rs:2151]: You can also remove this comment.

Comment (MunishMummadi) [reply components/layout_2020/flow/inline.rs:2151]: That's what I thought. Will remove.

**Comment by Loirooriol** [comment]
We first need to land a PR that only enables the tests without changing layout. That's not this nor #31913.

**Comment by MunishMummadi** [comment]
Got it. You want me to 
1.Remove the line `[css-inline]` skip: true from the `servo/tests/wpt/include.ini` file 
2.Not include any changes to the layout engine code related to the `baseline-source` implementation.

**Comment by Loirooriol** [comment]
Yes, you should post a PR that only modifies `servo/tests/wpt/include.ini` and updates test expectations, like

```
./mach build -r
./mach test-wpt -r --log-raw /tmp/servo.log tests/wpt/tests/css/css-inline/baseline-source/
./mach update-wpt /tmp/servo.log
./mach test-wpt -r --legacy-layout --log-raw /tmp/servo.log tests/wpt/tests/css/css-inline/baseline-source/
./mach update-wpt --legacy-layout /tmp/servo.log
git add tests/wpt/meta/css/css-inline/baseline-source/
git add tests/wpt/meta-legacy-layout/css/css-inline/baseline-source/
```

**Comment by MunishMummadi** [comment]
I will add a commit with those to the new PR . What do you think.

File: tests/wpt/include.ini [hunk 3]
```diff
@@ -49,6 +49,8 @@ skip: true
     skip: true
   [css-inline]
     skip: true
+  [css-inline/baseline-source]
+    skip: false
```
Comment (Loirooriol) [inline tests/wpt/include.ini:53]: You should now remove this, rebase on top of #31913, and update test expectations.

Comment (MunishMummadi) [reply tests/wpt/include.ini:53]: So, as the #31913 is merged. I will rebase my branch on top of `main`. I will be remove those line stated above. But in order to run tests. I need to have 
`  [css-inline]
    skip: true
    [baseline-source]
      skip: false` . Is my understanding correct??

Comment (Loirooriol) [reply tests/wpt/include.ini:53]: Yes I meant that the patch shouldn't change this file.

**Comment by Loirooriol** [comment]
I think you rebased main on top of your patch. You need to rebase your patch on top of main.

**Comment by MunishMummadi** [comment]
My vscode isn't showing the conflicts. I am editing them in vim. Sorry for the issue.

**Comment by Loirooriol** [comment]
Can you squash the commits together? E.g. 925f391fdccd19a1d4e0eb5ecce791f4bd2bb92e is broken

**Comment by MunishMummadi** [comment]
I squashed the last 2 commits. Is this what you are expecting. Correct me if I am wrong

File: components/layout_2020/flow/inline.rs (hunk 2 above)
Comment (Loirooriol) [reply components/layout_2020/flow/inline.rs:2151]: It's not removed.

**Review by Loirooriol — COMMENTED** [review]
I would in fact squash everything into a single commit, since e.g. it's a bit weird that d69f0cd5aac3b2e85eff33a01ead6f0aa4d317ab patches inline.rs instead of doing that as part of cf9b0cccf539da5830f4869b2e23429a45a37cca. But not a big deal

File: components/layout_2020/flow/inline.rs (hunk 2 above)
Comment (MunishMummadi) [reply components/layout_2020/flow/inline.rs]: It's my bad. I removed it. But while resolving conflicts. I accidently added it.

**Comment by Loirooriol** [comment]
I have landed https://github.com/servo/stylo/pull/26.
So can you update your PR to avoid modifying Cargo.toml, but using the updated Cargo.lock?

**Comment by MunishMummadi** [comment]
So, I have to run `cargo update` on my local branch and push the changes. Is my understanding correct?

**Comment by Loirooriol** [comment]
No, please only update the style crates. `cargo update -p style` or something.
It sould do the trick if you just rebase your patch to tip, run `git checkout main -- Cargo.toml`, and compile to update `Cargo.lock`.

**Comment by MunishMummadi** [comment]
understood. I am on it.

**Comment by MunishMummadi** [comment]
@Loirooriol I encountered 
```console
ellation)
error: encountered diff marker
  --> components/layout_2020/flow/inline.rs:87:1
   |
87 | <<<<<<< HEAD
   | ^^^^^^^ after this is the code before the merge
88 | use style::values::generics::font::LineHeight;
89 | =======
   | -------
...
92 | >>>>>>> d962c23bd1 (Implementation-with-tests)
   | ^^^^^^^ above this are the incoming code changes
   |
   = help: if you're having merge conflicts after pulling new code, the top section is the code you already had and the bottom section is the remote code
   = help: if you're in the middle of a rebase, the top section is the code being rebased onto and the bottom section is the code coming from the current commit being rebased
   = note: for an explanation on these markers from the `git` documentation, visit <https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_checking_out_conflicts>

error: could not compile `layout_2020` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
^Cmach interrupted by signal or user action. Stopping.pt
munish@Zoro:~/servo$ ./mach build
No build type specified so assuming `--dev`.
   Compiling script v0.0.1 (/home/munish/servo/components/script)
   Compiling layout_2020 v0.0.1 (/home/munish/servo/components/layout_2020)
error[E0432]: unresolved import `style::values::generics::text::LineHeight`
  --> components/layout_2020/flow/inline.rs:87:5
   |
87 | use style::values::generics::text::LineHeight;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `LineHeight` in `values::generics::text`
   |
help: consider importing one of these items instead
   |
87 | use style::properties::LonghandId::LineHeight;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
87 | use style::properties::PropertyDeclaration::LineHeight;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
87 | use style::properties::animated_properties::AnimationValue::LineHeight;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
87 | use style::values::computed::LineHeight;
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     and 2 other candidates

error[E0599]: no method named `clone_baseline_source` found for reference `&servo_arc::Arc<style::properties::ComputedValues>` in the current scope
    --> components/layout_2020/flow/inline.rs:2147:28
     |
2147 |         match self.style().clone_baseline_source() {
     |                            ^^^^^^^^^^^^^^^^^^^^^ help: there is a method with a similar name: `clone_border_image_source`

Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `layout_2020` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Comment by Loirooriol** [comment]
> use style::values::generics::text::LineHeight;

You didn't rebase to tip: https://github.com/servo/servo/blob/8c8d68e415bbea24593476d32f4bb7c46016bca3/components/layout_2020/flow/inline.rs#L87

> no method named `clone_baseline_source`

I guess `Cargo.lock` is not being updated.

**Comment by MunishMummadi** [comment]
Yeah `cargo.lock` is not getting updated. It is showing conflicts. Can I remove the conflicts manually.

**Comment by MunishMummadi** [comment]
I manually updated the `Cargo.toml` and `Cargo.lock` . But even then I am getting the 2nd error stated above 
```console
munish@Zoro:~/servo$ ./mach build
No build type specified so assuming `--dev`.
   Compiling layout_2020 v0.0.1 (/home/munish/servo/components/layout_2020)
error[E0599]: no method named `clone_baseline_source` found for reference `&servo_arc::Arc<style::properties::ComputedValues>` in the current scope
    --> components/layout_2020/flow/inline.rs:2148:28
     |
2148 |         match self.style().clone_baseline_source() {
     |                            ^^^^^^^^^^^^^^^^^^^^^ help: there is a method with a similar name: `clone_border_image_source`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `layout_2020` (lib) due to previous error
      Timing report saved to /home/munish/servo/target/cargo-timings/cargo-timing-20240331T234933Z.html
Failed in 0:00:02
[Warning] Could not generate notification: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.Notifications was not provided by any .service files
munish@Zoro:~/servo$
```

**Comment by Loirooriol** [comment]
Please force-push what you have to this PR.
Cargo.lock isn't supposed to be edited manually.

**Comment by MunishMummadi** [comment]
Sorry for my wording. I meant resolving conflicts manually in `Cargo.lock`

**Review by Loirooriol — CHANGES_REQUESTED** [review]
Please improve the commit message, this doesn't make sense:

```
Updated Cargo files

Implementation-with-tests

implemented-baseline-source

cargo-files

squahed-commits

Enable baseline-source tests in css-inline-minor code changes

wpt: Enable tests in `/css/css-inline/baseline-source/` (https://github.com/servo/servo/pull/31913)

* updated-baseline-source

* updated-for-tests-with-fmt

rebased-for-tests

minor-change
```

**Comment by MunishMummadi** [comment]
> Please improve the commit message, this doesn't make sense:
> 
> ```
> Updated Cargo files
> 
> Implementation-with-tests
> 
> implemented-baseline-source
> 
> cargo-files
> 
> squahed-commits
> 
> Enable baseline-source tests in css-inline-minor code changes
> 
> wpt: Enable tests in `/css/css-inline/baseline-source/` (https://github.com/servo/servo/pull/31913)
> 
> * updated-baseline-source
> 
> * updated-for-tests-with-fmt
> 
> rebased-for-tests
> 
> minor-change
> ```

It's an over-look . Sorry for that.

**Comment by Loirooriol** [comment]
I have edited the message to add the missing hyphen in `baseline-source`.
Also is you say "with tests" it seems you are adding new tests, but this is only updating test expectations.

**Review by Loirooriol — APPROVED** [review]

**Pull request merged**